Add new HomeSeer Z-Wave device types

kmitch95120 opened the pull request:
* Added support for the following HS devices as a subclass of HSSensor:

  - HSTemperature
  - HSHumidity
  - HSLuminance
  - HSFanState
  - HSOperatingState


* Created manifest.json file to by HA 0.92+ compliant.

* Added config option 'location2_names' and changed the functionality of 'location_names'.

* Changed entity attribute 'Location 2' to 'Location2' to align with config option.

* Added entity attributes "Location" and "Location2" to all platforms.

kmitch95120 committed:
* Created manifest.json file to by HA 0.92+ compliant.

* Added config option 'location2_names' and changed the functionality of 'location_names'.

* Changed entity attribute 'Location 2' to 'Location2' to align with config option.

* Added entity attributes "Location" and "Location2" to all platforms.

kmitch95120 committed:
* Added support for the following HS devices as a subclass of HSSensor:

  - HSTemperature
  - HSHumidity
  - HSLuminance
  - HSFanState
  - HSOperatingState

marthoc commented:
Please see issue #4. Your comments would be appreciated.

To speed up merging new device support we may wish to split that part of this PR into its own PR to keep it separate from naming convention changes.

kmitch95120 commented:
I will remove the following change from my PR: 

* Added config option 'location2_names' and changed the functionality of 'location_names'.

None of the other changes conflict with Issue #4.

kmitch95120 committed:
* Removed the 'location2_names' parameter.

marthoc commented on sensor.py:
```diff
@@ -106,3 +123,95 @@ def icon(self):
+class HSTemperature(HSSensor):
+    """Representation of a HomeSeer temperature sensor device."""
+
+    @property
+    def unit_of_measurement(self):
+        if self._units is not None:
+            if str(self._units) == 'imperial':
+                return "°F"
+            elif str(self._units) == 'metric':
+                return "°C"
+            else:
+                return "?"
+        else:
+            return '?'
```
What we need to figure out here is what unit the data comes from HomeSeer in - i.e. is value sent over the API in Celsius or Fahrenheit? Home Assistant will handle the conversion/representation in the frontend etc automatically depending on what localization option you select.

kmitch95120 replied:
The values coming from HS do not have units and there isn't anything in the API documentation that talks about a way to get the unit setting from HS.  I made the assumption that the unit system selected in Home Assistant would likely match the unit system selected in HomeSeer.

marthoc replied:
I guess my question is, are the values from the API always in the same unit, or do they change depending on what the user has selected in HomeSeer. Ecobee, for example, always send Fahrenheit from their API regardless of what localization settings are selected for the account. Makes it easier to automate. 

If the value over the HS API is not consistent, then the way to solve this is through a configuration.yaml config key. We should not make assumptions for the user.

marthoc commented on sensor.py:
```diff
@@ -106,3 +123,95 @@ def icon(self):
+        else:
+            return '?'
+
+    @property
+    def icon(self):
+        if self._units is not None:
+            if str(self._units) == 'imperial':
+                if self.state < 50:
+                    return 'mdi:thermomter-low'
+                elif self.state > 90:
+                    return 'mdi:thermometer-high'
+                else:
+                    return 'mdi:thermometer'
+            elif str(self._units) == 'metric':
+                if self.state < 10:
+                    return 'mdi:thermomter-low'
+                elif self.state > 33:
+                    return 'mdi:thermometer-high'
+                else:
+                    return 'mdi:thermometer'
+            else:
+                return 'mdi:thermometer'
+        else:
+            return 'mdi: thermometer'
```
Same as above - if we know what unit the data is coming from HomeSeer in, then we can figure out what the icons should be always. Knowing what unit Home Assistant is using doesn't help us here.

kmitch95120 replied:
See above.  Is it not "safe" to assume that the unit system set in HA is the same as what would have been set in HS?

marthoc commented on sensor.py:
```diff
@@ -106,3 +123,95 @@ def icon(self):
+        return DEVICE_CLASS_ILLUMINANCE
+
+
+class HSFanState(HSSensor):
+    """Representation of a HomeSeer HVAC fan state sensor device."""
+
+    @property
+    def icon(self):
+        if self.state == 0:
+            return 'mdi:fan-off'
+        else:
+            return 'mdi:fan'
```
If this is simply on/off sensor, it should be a binary sensor rather than a sensor.

kmitch95120 replied:
No. It can be a value in the range of 0 - 8, with 0 being 'off' and 1-8 being some state of 'on'.  I elected not to cover all 'on' states just for an icon display.

marthoc commented on sensor.py:
```diff
@@ -106,3 +123,95 @@ def icon(self):
+class HSOperatingState(HSSensor):
+    """Representation of a HomeSeer HVAC operating state sensor device."""
+
+    @property
+    def icon(self):
+        if self.state == 0:         # Idle
+            return 'mdi:fan-off'
+        elif self.state == 1:       # Heating
+            return 'mdi:flame'
+        elif self.state == 2:       # Cooling
+            return 'mdi:snowflake'
+        else:                       # Fan Only
+            return 'mdi:fan'
```
We should override the state property of the base class to translate the numeric values to strings like 'idle', 'heat', 'cool', and 'fan' for a better representation in the frontend.

kmitch95120 replied:
I will add that.

kmitch95120 replied:
I think this should be done for 'fan state' as well even though the icon is only off or on. 

thoughts?

marthoc replied:
Yes what are the values of fan state? 0 means off, 8 means high, other values are in between? As long as the representation in HS is consistent, then we should use the strings rather than numeric values.

marthoc commented on sensor.py:
```diff
@@ -34,9 +48,10 @@ async def async_setup_platform(hass, config, async_add_entities,
 class HSSensor(Entity):
     """Representation of a HomeSeer sensor-type device."""
-    def __init__(self, device, connection):
+    def __init__(self, device, connection, units=None):
```
See my other comments re temperature, this additional keyword parameter will not be required.

kmitch95120 replied:
This will be true if we can figure out a way to get the unit system from HS but I can't find any way to do that using the API.

marthoc replied:
Looking at the API I see this request: 

`/JSON?request=getsetting&setting=SETTING_NAME`

The documentation gives the example:

```
 Returns the value for a specific settting. For example, the setting for the name of location 1 is called "gLocLabel". To get the name of this label use:

/JSON?request=getsetting&setting=gLocLabel

The return might be:

{
"Value":"Room"
}
```

Possibly the localization/temperature unit setting could be obtained this way?

kmitch95120 replied:
I don't see anything in settings.ini that indicates the temperature scale setting.   Taking a look around some other files.

kmitch95120 replied:
I did some searching on the HS forums and this has been an ongoing problem for various setups.  Many of the devices have their own scale setting as a part of their protocol while others rely on the HS setting.  For example, the TEMP05 1-wire logger has a command to set the scale for its readings and also an option to append "F" or "C" to the end of the readings.  When sending these readings to HA via MQTT using device_string instead of device_value, it's easy to determine the units based on the char at the end of the string. 

I'm not even sure a configuration key would solve this since readings could be in either scale depending upon how the device or devices were configured on the HS side.  I discovered this with light sensors as well.  Some of my sensors report in percentage from 0-100% while others report in Lux from 0 - ~2500.  

I'm wondering if it would be better not to handle this in the component at all and set the units using entity customization. I'm new to HA so still learning.

marthoc replied:
I think you might be right here - given there's no consistency, we should just report the value without units or icon modifications, and let the user customize. No real need for a HSTemperature class in that case either.

marthoc replied:
Another question - since this component only handles ZWave devices (and I would want to keep it that way) is temp reporting part of the ZWave spec at all? I would think that maybe all ZWave devices report in Fahrenheit?

kmitch95120 replied:
Yes, It's a part of the HRV Status Command Class.  Section 4.45, specifically 4.45.2 HRV Status Report Command.

marthoc commented on sensor.py:
```diff
@@ -15,15 +16,28 @@
 
     for device in homeseer.devices:
         if device.device_type_string in HASS_SENSORS:
             if device.device_type_string == DEVICE_ZWAVE_BATTERY:
                 dev = HSBattery(device, homeseer)
+            elif device.device_type_string == DEVICE_ZWAVE_TEMPERATURE:
+                dev = HSTemperature(device, homeseer, units)
+            elif device.device_type_string == DEVICE_ZWAVE_RELATIVE_HUMIDITY:
+                dev = HSHumidity(device, homeseer)
+            elif device.device_type_string == DEVICE_ZWAVE_LUMINANCE:
+                dev = HSLuminance(device, homeseer)
+            elif device.device_type_string == DEVICE_ZWAVE_FAN_STATE:
+                dev = HSFanState(device, homeseer)
+            elif device.device_type_string == DEVICE_ZWAVE_OPERATING_STATE:
+                dev = HSOperatingState(device, homeseer)
             else:
                 dev = HSSensor(device, homeseer)
```
To clean this up we should create a factory method similar to the get_zwave_device in pyhs3 that takes the string as input and returns the appropriate object.

kmitch95120 replied:
I will make that change.

marthoc commented:
@kmitch95120 I updated the base branch to 0.4 since this will be part of the next release. In addition to making the changes discussed above you will need to resolve the merge conflicts that have been introduced. Thanks!

kmitch95120 commented:
I'll open a new pull request once I get my repo straightened out.
